dll/ntdll/rtl/version.c

```diff
@@ -89,14 +90,110 @@ SetRosSpecificInfo(IN OUT PRTL_OSVERSIONINFOEXW VersionInformation)
  *
  * REVISIONS
  *  2000-08-10 ekohl
+ *  2020-03-03 Fraizeraust
  *
  * @implemented
  */
 BOOLEAN NTAPI
 RtlGetNtProductType(PNT_PRODUCT_TYPE ProductType)
 {
-    *ProductType = SharedUserData->NtProductType;
-    return TRUE;
+    HANDLE hKey;
+    ULONG ReturnLength;
+    OBJECT_ATTRIBUTES ObjectAttributes;
+    NTSTATUS Status;
+    UCHAR Buffer[sizeof(PKEY_VALUE_PARTIAL_INFORMATION) + MAX_PATH * sizeof(WCHAR)];
+    PKEY_VALUE_PARTIAL_INFORMATION PartialInformation = (PVOID)Buffer;
+    UNICODE_STRING KeyProduct = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\SYSTEM\\CurrentControlSet\\Control\\ProductOptions");
+    UNICODE_STRING ValueNtProduct = RTL_CONSTANT_STRING(L"ProductType");
+    UNICODE_STRING WorkstationProduct;
+    UNICODE_STRING LanManProduct;
+    UNICODE_STRING NtProductType;
+
+    /* Do a sanity check -- validate our parameter first */
+    if (!ProductType)
+    {
+        DPRINT1("RtlGetNtProductType(): An invalid NT_PRODUCT_TYPE variable pointer was given!\n");
+        *ProductType = NtProductWinNt;
```

Comment:
Invalid ... yet assigned to?

Reply:
http://www.codewarrior.cn/ntdoc/winnt/rtl/RtlGetNtProductType.htm

Reply:
Yes but then this means that the function just expects the pointer to always be valid.
Can you try in your test to call `RtlGetNtProductType(NULL)` and see what it does?

Reply:
> RtlGetNtProductType.htm

So what?

> test

Reminder: I already gave the results in #2400.